src/compiler/checker.ts

```diff
@@ -36065,8 +36065,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const declaration = parameter.valueDeclaration as ParameterDeclaration;
             if (!getEffectiveTypeAnnotationNode(declaration)) {
                 let type = tryGetTypeAtPosition(context, i);
-                if (type && type.flags & TypeFlags.Never && declaration.initializer) {
-                    type = widenTypeInferredFromInitializer(declaration, checkDeclarationInitializer(declaration, CheckMode.Normal));
+                if (type && declaration.initializer) {
+                    const initializerType = widenTypeInferredFromInitializer(declaration, checkDeclarationInitializer(declaration, CheckMode.Normal));
+                    if (!isTypeAssignableTo(initializerType, type) && isTypeAssignableTo(type, initializerType)) {

```

Comment:
Hmm... what happens in the case of mutually assignable types?

Reply:
How many of them do we have? There is, of course, `any` and I can also think of bivariant methods but I can't call more on the spot.

That said - if the types are mutually assignable then the old behavior is preserver (the contextual parameter type is preferred). Can you think of any scenario in which this would be problematic?

Reply:
I guess you’re right.  If the types are mutually assignable then it probably doesn’t matter which one is picked and in the absence of a type annotation it makes more sense to pick the contextual type (especially since it prevents inferred `any`s from sneaking in)

Reply:
> especially since it prevents inferred anys from sneaking in

That I'm not so sure about :p

```ts
const test: (arg: any) => void = (arg = 5) => {
  arg
  // ^? (parameter) arg: any
};
```

But inferring `number` would also not be good here either. It would only give the illusion of type-safety within the function while `test` would still be callable with just anything.

Reply:
I specifically meant inferring `any` from a default when there’s a more sane contextual type available.